src/gsy_e/models/strategy/heat_pump_soc_management.py

```diff
@@ -0,0 +1,137 @@
+
+    def __init__(self, energy_params: "HeatPumpEnergyParameters"):
+        self._last_switch: Optional[DateTime] = None
+        self._energy_params = energy_params
+        self._charger = energy_params.combined_state._charger
+        self._current_state = HeatPumpChargingState.CHARGE
+        if isinstance(GlobalConfig.MARKET_MAKER_RATE, dict):
+            self._average_rate = mean(GlobalConfig.MARKET_MAKER_RATE.values())
+        else:
+            assert False
+
+    def calculate(self, time_slot: DateTime, _buy_rate: float = 0.0):
+        """
+        Calculate the bid energy depending on the current state of the heat pump, the current SOC
+        and whether the market maker rate is cheap or expensive compared to the average market
+        maker rate.
+        """
+        target_state = self._current_state
+        if not self._is_time_for_state_change(time_slot):
+            # If state change is not possible, but at the same time the soc is below the min or
+            # above the max SOC value of the tank, then maintain the SOC.
+            if self._charger.get_average_soc(time_slot) >= self.MAX_SOC_TOLERANCE:
+                target_state = HeatPumpChargingState.MAINTAIN_SOC
+            if self._charger.get_average_soc(time_slot) <= self.MIN_SOC_TOLERANCE:
+                target_state = HeatPumpChargingState.MAINTAIN_SOC
+        else:
+            # If the state change is possible, check the market maker rate to set the new state
+            target_state = self._should_charge_or_discharge(time_slot)
+            if self._current_state != target_state:
+                # If the target state is the same as the current state, do nothing. Otherwise,
+                # update the current state and the last switch timestamp.
+                self._current_state = target_state
+                self._last_switch = time_slot
+
+        return self._get_energy_from_target_state(target_state, time_slot)
+
+    def _is_time_for_state_change(self, time_slot: DateTime) -> bool:
+        if not self._last_switch:
+            # If the last_switch has not been set yet, the simulation is starting and no switch has
+            # occurred yet. Change state if needed.
+            self._last_switch = time_slot
+            return True
+        if time_slot - self._last_switch < duration(minutes=self.MINUTES_BEFORE_SWITCH_ALLOWED):
+            # If not enough time has passed since the last switch, do not allow state change.
+            return False
+        # Otherwise, allow the state change.
+        return True
+
+    def _get_energy_from_target_state(
+        self, target_state: HeatPumpChargingState, time_slot: DateTime
+    ) -> float:
+        if target_state == HeatPumpChargingState.CHARGE:
+            return self._energy_params.get_max_energy_demand_kWh(time_slot)
+        if target_state == HeatPumpChargingState.DISCHARGE:
+            return 0
+        return self._energy_params.get_min_energy_demand_kWh(time_slot)
```

Comment:
This does not work unfortunately. If the heat demand is higher than the current charge, the storage will be "over-drained". We need to return the current stored energy energy ("soc_energy") here. The ``
already deals with this actually both in the [PCM tank state](https://github.com/gridsingularity/gsy-e/blob/a66e67fd92a64bbfcdf6f34f1f9966a1c26c6fac/src/gsy_e/models/strategy/state/heatpump_tank_states/pcm_tank_state.py#L256) and in the [water tank state](https://github.com/gridsingularity/gsy-e/blob/a66e67fd92a64bbfcdf6f34f1f9966a1c26c6fac/src/gsy_e/models/strategy/state/heatpump_tank_states/water_tank_state.py#L138).
So my proposal is the following:
```suggestion
        if target_state == HeatPumpChargingState.CHARGE:
            return self._energy_params.get_max_energy_demand_kWh(time_slot)
        return self._energy_params.get_min_energy_demand_kWh(time_slot)
```

Reply:
We somehow need to divide the responsibilities between the states and this SOC management.

Reply:
Unfortunately, the SOC Management is only applied before trading. 
The actual charging and discharging happens in the energy parameters of the heat pump [here](https://github.com/gridsingularity/gsy-e/blob/a66e67fd92a64bbfcdf6f34f1f9966a1c26c6fac/src/gsy_e/models/strategy/energy_parameters/heatpump/heat_pump.py#L215)

If there is no energy traded, the tank indeed could over-discharge if the heat demand is higher that the SOC of the tank.

Reply:
We discussed offline, that this approach is correct, because the heat demand of one market slot should not exceed the `MIN_SOC_TOLERANCE`. The latter should be configured accordingly.

Reply:
Adapted, thanks!